### PR TITLE
Rewrite tests

### PR DESCRIPTION
**Pull request by VladimirGr**

### CODE REVIEW

File: tests/hooks/index.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export { useIsFirstRender } from './useIsFirstRender';
+
+export { useWhenIsNotFirstRender } from './useWhenIsNotFirstRender';
```
Comment (piskunovim) [inline tests/hooks/index.ts:3]: Why?

File: tests/hooks/index.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export { useIsFirstRender } from './useIsFirstRender';
```
Comment (piskunovim) [inline tests/hooks/index.ts:1]: Not sure, about isFirstRender/WhenIsNotFirstRender here. What do you think about:

```suggestion
import { useEffect, useRef } from "react";

// only first render
export const useDidMount = (callback: VoidFunction | (() => VoidFunction)) => {
  useEffect(callback, []);
};

// only on change
export const useDidUpdate = (
  callback: VoidFunction | (() => VoidFunction),
  dependencies: any[],
) => {
  const mountRef = useRef(false);

  useEffect(() => {
    if (!mountRef.current) {
      mountRef.current = true;
      return;
    }

    return callback();
  }, dependencies);
};
```

File: tests/useMemoReducer/useMemoReducer.test.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    fireEvent.click(incrementButton);
+    fireEvent.click(incrementButton);
+
+    expect(sideEffectRenderCount.textContent).toBe('0');
+    expect(countValue.textContent).toBe('2');
+    expect(objectValue.textContent).toBe(JSON.stringify({ value: 'Some value' }));
+  });
+
+  it('shouldn`t re-render a component when own properties weren`t changed', () => {
```
Comment (piskunovim) [inline tests/useMemoReducer/useMemoReducer.test.tsx:73]: It seems previous test covers this case

File: tests/useMemoReducer/defaultComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        onClick={() => dispatch({ type: Action.OBJECT_VALUE_CHANGED, value: 'Some value changed' })}
+      >
+        Change object
+      </button>
+    </div>
+  );
+};
+
+const TestComponentWithoutRenders: FC<Props> = ({ reducer = defaultReducer, initialState = defaultInitialState }) => {
+  const [sideEffectRender, updateSideEffectRender] = useState(0);
```
Comment (piskunovim) [inline tests/useMemoReducer/defaultComponent.tsx:58]: should it be useRef()?

File: tests/useMemoReducer/defaultComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  const isFirstRender = useRef(true);
+  const [useSelector, dispatch] = useMemoReducer(reducer, initialState);
+
+  const obj = useSelector((state) => state.object);
+
+  useEffect(() => {
+    if (!isFirstRender.current) {
```
Comment (piskunovim) [inline tests/useMemoReducer/defaultComponent.tsx:65]: What's the reason for not counting a mount as a render?

File: tests/useMemoReducer/defaultComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  reducer?: (state: State, action: Actions) => State;
+  initialState?: State;
+};
+
+const DefaultTestComponent: FC<Props> = ({ reducer = defaultReducer, initialState = defaultInitialState }) => {
+  const [sideEffectRender, updateSideEffectRender] = useState(0);
+  const isFirstRender = useRef(true);
+  const [useSelector, dispatch] = useMemoReducer(reducer, initialState);
+
+  const count = useSelector((state) => state.count);
```
Comment (piskunovim) [inline tests/useMemoReducer/defaultComponent.tsx:23]: It seems reasonable to create a <CountComponent /> for checking re-renders instead of making two similar components.

File: tests/useMemoReducer/useMemoReducer.test.tsx [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    const { getByTestId } = render(<TestComponentWithoutRenders />);
```
Comment (piskunovim) [inline tests/useMemoReducer/useMemoReducer.test.tsx:74]: I see the difference between the components, but if the re-render didn't happen when there was a count on the page, why do we expect it to happen when there isn't one? It seems previous test covers this case completely.

File: tests/thunks/thunks.test.tsx [hunk 8]
```diff
@@ -35,74 +17,70 @@ describe('thunks', () => {
...
-
-    fetchMock.mockResolvedValueOnce({ json: () => Promise.resolve({ title: 'Data 2' }) });
-
-    await act(async () => {
-      cleanup(); // Unmount the component
-      render(<ThunksTestComponent />);
-      await Promise.resolve();
-    });
+    const changedTitle = getCurrentHookValue(renderHook(() => useSelector((state) => state.title)));
+    const changedErrorMessage = getCurrentHookValue(renderHook(() => useSelector((state) => state.errorMessage)));
```
Comment (piskunovim) [inline tests/thunks/thunks.test.tsx:45]: This test is similar to the previous one, but checks for an exception handling path that is not directly related to the package (implementation of the Thunk itself). It might make sense to test the exception handling at the package level.

File: tests/useMemoReducer/defaultComponent.tsx (hunk 4 above)
Comment (VladimirGr) [reply tests/useMemoReducer/defaultComponent.tsx]: We don`t get an actual number of the re-renders in the DOM, because useRef doesn`t update a component

File: tests/useMemoReducer/defaultComponent.tsx (hunk 5 above)
Comment (VladimirGr) [reply tests/useMemoReducer/defaultComponent.tsx]: This question is not actual, because I rewrote tests

File: tests/useMemoReducer/defaultComponent.tsx (hunk 6 above)
Comment (VladimirGr) [reply tests/useMemoReducer/defaultComponent.tsx]: I didn't quite understand this suggestion. And I rewrote the test

File: tests/useMemoReducer/defaultComponent.tsx (hunk 4 above)
Comment (piskunovim) [reply tests/useMemoReducer/defaultComponent.tsx]: Yes, and ...?  Testing  useMemoReducer() subscriber logic, not useState() correct? If yes, then useRef is better choice.

File: tests/useMemoReducer/defaultComponent.tsx (hunk 6 above)
Comment (piskunovim) [reply tests/useMemoReducer/defaultComponent.tsx]: The suggestion part was missed: I suggested to make one instead two.

File: tests/hooks/useWhenIsNotFirstRender.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import { DependencyList, EffectCallback, useEffect, useRef } from 'react';
+
+import { useIsFirstRender } from './useIsFirstRender';
+
+export const useWhenIsNotFirstRender = (callback: EffectCallback, depends: DependencyList) => {
+  const isFirstRender = useIsFirstRender();
+
+  const callbackRef = useRef<EffectCallback | null>(null);
+
+  callbackRef.current = callback;
```
Comment (piskunovim) [inline tests/hooks/useWhenIsNotFirstRender.ts:10]: It seems not necessary to store the callback in a ref. The callback will not change during the execution, so it can be accessed directly.

File: tests/hooks/useWhenIsNotFirstRender.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import { DependencyList, EffectCallback, useEffect, useRef } from 'react';
+
+import { useIsFirstRender } from './useIsFirstRender';
+
+export const useWhenIsNotFirstRender = (callback: EffectCallback, depends: DependencyList) => {
```
Comment (piskunovim) [inline tests/hooks/useWhenIsNotFirstRender.ts]: Will try to reference this: https://github.com/piskunovim/useMemoReducer/pull/9/files/80342bdd79b39c78a3a6c8acf5c5081e670ea024#r1260154167
Should work good and increase readability.

File: tests/useMemoReducer/SimleComponent.tsx [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+export const DECREMENT_SELECTOR = 'decrement';
+export const CHANGE_OBJECT_SELECTOR = 'change-object';
+export const OBJECT_VALUE_SELECTOR = 'object-value';
+
+type Props = {
+  reducer?: (state: State, action: Actions) => State;
+  initialState?: State;
+};
+
+export const SimpleComponent: FC<Props> = ({ reducer = defaultReducer, initialState = defaultInitialState }) => {
```
Comment (piskunovim) [inline tests/useMemoReducer/SimleComponent.tsx:20]: Minor: filename SimleComponent -> SimpleComponent

File: tests/useMemoReducer/SimleComponent.tsx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  const [currentRenders, updateCurrentRenders] = useState(0);
+
+  const [allRenders, updateAllRenders] = useState(0);
```
Comment (piskunovim) [inline tests/useMemoReducer/SimleComponent.tsx:23]: I'm afraid I'm having trouble understanding the concept of incrementing the rerender count by causing additional rerenders. 

I think only **useMemoReducer()** should kick off rerenders. And the counter that keeps track of rerenders shouldn't be reactive. Right?

File: tests/useMemoReducer/defaultComponent.tsx (hunk 4 above)
Comment (piskunovim) [reply tests/useMemoReducer/defaultComponent.tsx]: That's right. I will try cover it in different comment

File: tests/useMemoReducer/SimleComponent.tsx (hunk 12 above)
Comment (piskunovim) [inline tests/useMemoReducer/SimleComponent.tsx:23]: I'm a bit confused about the idea of counting rerenders by causing more rerenders. I think only useMemoReducer() should kick off rerenders. And the counter that keeps track of rerenders shouldn't be reactive. Right?

Here is an example of what I mean, but imagine **useMemoReducer** instead the **useState**:
https://codesandbox.io/s/distracted-heisenberg-zh252q?file=/src/App.js:176-184

File: tests/useMemoReducer/useMemoReducer.test.tsx [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    const initialAllRenders = allRendersCount.textContent;
+
+    expect(initialAllRenders).toBe('1');
+
+    fireEvent.click(incrementButton);
+    fireEvent.click(incrementButton);
+    fireEvent.click(incrementButton);
+    fireEvent.click(decrementButton);
+
+    expect(currentRendersCount.textContent).toBe('0');
```
Comment (piskunovim) [inline tests/useMemoReducer/useMemoReducer.test.tsx:78]: The variable name 'currentRendersCount' isn't as clear as 'allRendersCount'. May be "objectRendersCount"?

File: tests/hooks/useWhenIsNotFirstRender.ts (hunk 9 above)
Comment (VladimirGr) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: We mustn`t lie to React. If we use functions and variables that exists in a Reactor data flow then we must add ones in hook depends. `callback` exists in the React data flow/

File: tests/hooks/index.ts (hunk 2 above)
Comment (VladimirGr) [reply tests/hooks/index.ts:1]: I want to have a separated hooks (userIsFirstRender, useWhenIsNotFirstRender). Now I don`t need for a useDidMount hook, because we don`t use it. I think that the name `useDidUpdate` does not reflect that the hook should be called ignoring the first render

File: tests/hooks/useWhenIsNotFirstRender.ts (hunk 10 above)
Comment (VladimirGr) [reply tests/hooks/useWhenIsNotFirstRender.ts]: Answered

File: tests/useMemoReducer/SimleComponent.tsx (hunk 11 above)
Comment (VladimirGr) [reply tests/useMemoReducer/SimleComponent.tsx:20]: Thank you!

Comment (VladimirGr) [reply tests/useMemoReducer/SimleComponent.tsx]: Fixed

File: tests/useMemoReducer/SimleComponent.tsx (hunk 12 above)
Comment (VladimirGr) [reply tests/useMemoReducer/SimleComponent.tsx]: Yes, so I use this code:
```js
  useEffect(() => {
    updateAllRenders((renders) => renders + 1);
  }, [obj, count]);

  useWhenIsNotFirstRender(() => {
    updateCurrentRenders((renders) => renders + 1);
  }, [obj]);
```
We can see that allRenders and currentRenders will be update when a state of the useMemoReducer was updated

File: tests/useMemoReducer/useMemoReducer.test.tsx (hunk 13 above)
Comment (VladimirGr) [reply tests/useMemoReducer/useMemoReducer.test.tsx:78]: Perhaps in other tests we`ll need to check a currentRenders when not only the object was changed.

File: tests/hooks/index.ts (hunk 2 above)
Comment (piskunovim) [reply tests/hooks/index.ts:1]: Why? **useDidRender** won't be triggered first render.

File: tests/hooks/useWhenIsNotFirstRender.ts (hunk 9 above)
Comment (piskunovim) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: **callback** is a standard function passed to useEffect dispatcher as a **create()** function. There is no way to lie React here. React allows to pass callbacks this way.

I see what you're trying to prevent, but I say it's unnecessary.

Comment (piskunovim) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: It's internal code. No one will use it differently.

File: tests/hooks/index.ts (hunk 2 above)
Comment (VladimirGr) [reply tests/hooks/index.ts:1]: Why do you say about the `useDidRender` hook? There was no anything about one in the previous comment, you suggested using the `useDidUpdate ` hook.

Comment (piskunovim) [reply tests/hooks/index.ts:1]: Right, **useDidUpdate**. Mistyped in the previous comment. Please checkout  implementation of the **useDidUpdate**.

File: tests/hooks/useWhenIsNotFirstRender.ts (hunk 9 above)
Comment (VladimirGr) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: You are right `callback` is a standard `create` function for the React dispatcher. But we write a custom hook, so in a hook  context our `callback` is an ordinary dependency and we need to add one to the dependencies list of the hook.

if we don`t, do it, we`ll get a warning:

image

Comment (piskunovim) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: I have no issue with dependency. I have concern about unnecessary callback Ref, as I keep saying.

Comment (piskunovim) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: It's obvious that we need to pass callback as dependency, but why we need ref?

Comment (VladimirGr) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: If you're interested, you can read this tread where Dan Abramov wrote about it.

Dan abramov – developer on Facebook core – [strongly recommend to keep that rule enabled](https://github.com/facebook/create-react-app/issues/6880#issuecomment-485912528).

Comment (VladimirGr) [reply tests/hooks/useWhenIsNotFirstRender.ts:10]: If we use the React hooks, I think we should follow their rules. It will be the right thing to do.

File: tests/hooks/index.ts (hunk 2 above)
Comment (VladimirGr) [reply tests/hooks/index.ts]: fixed it

File: tests/thunks/thunks.test.tsx (hunk 8 above)
Comment (VladimirGr) [reply tests/thunks/thunks.test.tsx:45]: I think, we don`t need to do this tests at all :)

Comment (VladimirGr) [reply tests/thunks/thunks.test.tsx:45]: I think, we don`t need to do this tests at all :) They don`t have any sense.
